### PR TITLE
Archana

### LINKED ISSUE #9: Implement OmegaCore's `intervene` interface
The interface to interventions in Omega is random-variable centric.

That is, you construct interventions in the form:

`Y_x = intervene(Y, X => x)`

where `Y_x` and `X` are random variables.

Currently #8 implements interventions at the model level, so we need to bridge between the two

### LINKED ISSUE #16: Use Omega to perform inference to compute probability terms
See CF paper (page 8)

### LINKED ISSUE #20: Implement FairLearning algorithm from CF paper

### LINKED ISSUE #21: Add method to compute path-specific counterfactuals
Assuming identifiability of path specific effects, compute the counterfactuals given blocked paths.

### PR DESCRIPTION
**Pull request by archanarw**

### CODE REVIEW

File: src/path_specific.jl [hunk 1]
```diff
@@ -0,0 +1,32 @@
+using Omega, Random
+import Omega:intervene, Interventions
+
+export apply_ps_intervention, BlockedEdges
```
Comment (mschauer) [inline src/path_specific.jl]: `BlockedEdges` is this defined?

File: src/prob_causal_graph.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    `empirical_mechanism((; μ=..., Σ=...), A, B)`
+ Emprical mechanism of A given B
+"""
+function empirical_mechanism(P, A, B)
+    Z = P.Σ[A,B] * inv(P.Σ[B,B])
+    # (; μ = P.μ[A] + Z*(xB - P.μ[B]), Σ = P.Σ[A,A] - Z*P.Σ[B,A])
+    (; a=Z, b=P.μ[A] + Z * ( - P.μ[B]), c=P.Σ[A,A] - Z * P.Σ[B,A])
+end
+
+function prob_causal_graph(df; p=0.01, test=gausscitest)
```
Comment (mschauer) [inline src/prob_causal_graph.jl]:
```suggestion
function prob_causal_graph(df, cg; p=0.01, test=gausscitest)
```
Better make `cg` an argument and provide a second method which calls it using `cg = pcalg(df, p, test)`

File: src/path_specific.jl (hunk 1 above)
Comment (archanarw) [reply src/path_specific.jl]: No, not yet. I meant to make it a struct, I'll do it now.

File: src/causalmodel.jl [hunk 3]
```diff
@@ -69,39 +56,67 @@ rem_edge!(g::CausalModel, x...) = rem_edge!(g.dag, x...)
 indegree(g::CausalModel) = indegree(g.dag)
 outdegree(g::CausalModel) = outdegree(g.dag)
 
+"""
+Defines the variable `varname` in the model
+"""
 struct CausalVar
     model::CausalModel
     varname::Symbol
 end
```
Comment (mschauer) [inline src/causalmodel.jl]:
```suggestion
struct CausalVar{T}
    model::CausalModel{T}
    varname::Symbol
end
```

File: src/path_specific.jl [hunk 4]
```diff
@@ -1,32 +1,52 @@
 using Omega, Random
+using LightGraphs:SimpleEdge
+
 import Omega:intervene, Interventions
 
 export apply_ps_intervention, BlockedEdges
 
-function apply_ps_intervention(model::CausalModel, x₁::Intervention, blocked_edges::AbstractVector{Bool}, x₂::Intervention, ω::AbstractΩ)
+struct BlockedEdges
+    edges::Vector{Edge}
```
Comment (mschauer) [inline src/path_specific.jl]: `edges` should also be parametrised by a concrete type. Ref: https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container

File: src/prob_causal_graph.jl [hunk 5]
```diff
@@ -12,8 +12,7 @@ function empirical_mechanism(P, A, B)
     (; a=Z, b=P.μ[A] + Z * ( - P.μ[B]), c=P.Σ[A,A] - Z * P.Σ[B,A])
 end
 
-function prob_causal_graph(df; p=0.01, test=gausscitest)
-    @show cg = pcalg(df, p, test)
+function prob_causal_graph(df, cg; p=0.01, test=gausscitest)
```
Comment (mschauer) [inline src/prob_causal_graph.jl]:
```suggestion
function prob_causal_graph(df, cg)
```
I think not needed anymore?

File: src/prob_causal_graph.jl [hunk 6]
```diff
@@ -12,6 +12,7 @@ function empirical_mechanism(P, A, B)
     (; a=Z, b=P.μ[A] + Z * ( - P.μ[B]), c=P.Σ[A,A] - Z * P.Σ[B,A])
 end
 
+"Returns the fully specified causal graph assuming gausain mechanism for each variable"
```
Comment (mschauer) [inline src/prob_causal_graph.jl:15]:
```suggestion
"Returns the fully specified causal graph assuming Gaussian mechanism for each variable"
```

**Pull request closed without merging**